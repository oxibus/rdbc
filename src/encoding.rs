// https://github.com/hsivonen/recode_rs

use std::io::{Read, Write};
use std::str::from_utf8_mut;

use encoding_rs::*;

use crate::error::DbcError;

pub fn utf8_to_gbk(src_data: &[u8]) -> Result<Vec<u8>, DbcError> {
    recode(src_data, "UTF-8", "GBK")
}

pub fn gbk_to_utf8(src_data: &[u8]) -> Result<Vec<u8>, DbcError> {
    recode(src_data, "GBK", "UTF-8")
}

pub fn to_utf8(src_encoding_label: &str, src_data: &[u8]) -> Result<Vec<u8>, DbcError> {
    recode(src_data, src_encoding_label, "UTF-8")
}

pub fn recode(
    src_data: &[u8],
    src_encoding_label: &str,
    dst_encoding_label: &str,
) -> Result<Vec<u8>, DbcError> {
    let src_encoding = get_encoding(Some(src_encoding_label.to_string()))?;
    let dst_encoding = get_encoding(Some(dst_encoding_label.to_string()))?;

    let mut decoder = src_encoding.new_decoder();
    let mut encoder = dst_encoding.new_encoder();

    let mut buf = std::io::Cursor::new(Vec::new());

    convert_via_utf8(
        &mut decoder,
        &mut encoder,
        &mut std::io::Cursor::new(src_data),
        &mut buf,
        true,
    )?;

    Ok(buf.into_inner())
}

pub fn get_encoding(opt: Option<String>) -> Result<&'static Encoding, DbcError> {
    match opt {
        None => Ok(UTF_8),
        Some(label) => match Encoding::for_label(label.as_bytes()) {
            None => Err(DbcError::InvalidEncodingLabel(label)),
            Some(encoding) => Ok(encoding),
        },
    }
}

pub fn convert_via_utf8(
    decoder: &mut Decoder,
    encoder: &mut Encoder,
    read: &mut dyn Read,
    write: &mut dyn Write,
    last: bool,
) -> Result<(), DbcError> {
    let mut input_buffer = [0u8; 2048];
    let mut intermediate_buffer_bytes = [0u8; 4096];
    let intermediate_buffer: &mut str = from_utf8_mut(&mut intermediate_buffer_bytes[..]).unwrap();
    let mut output_buffer = [0u8; 4096];
    let mut current_input_ended = false;
    while !current_input_ended {
        match read.read(&mut input_buffer) {
            Err(e) => {
                log::error!("Error reading input, error = {}", e);
                return Err(DbcError::EncodingReadInputError);
            }
            Ok(decoder_input_end) => {
                current_input_ended = decoder_input_end == 0;
                let input_ended = last && current_input_ended;
                let mut decoder_input_start = 0usize;
                loop {
                    let (decoder_result, decoder_read, decoder_written, _) = decoder.decode_to_str(
                        &input_buffer[decoder_input_start..decoder_input_end],
                        intermediate_buffer,
                        input_ended,
                    );
                    decoder_input_start += decoder_read;

                    let last_output = if input_ended {
                        match decoder_result {
                            CoderResult::InputEmpty => true,
                            CoderResult::OutputFull => false,
                        }
                    } else {
                        false
                    };

                    // Regardless of whether the intermediate buffer got full
                    // or the input buffer was exhausted, let's process what's
                    // in the intermediate buffer.

                    if encoder.encoding() == UTF_8 {
                        // If the target is UTF-8, optimize out the encoder.
                        if let Err(e) =
                            write.write_all(&intermediate_buffer.as_bytes()[..decoder_written])
                        {
                            log::error!("Error writing output, error = {}", e);
                            return Err(DbcError::EncodingWriteOutputError);
                        }
                    } else {
                        let mut encoder_input_start = 0usize;
                        loop {
                            let (encoder_result, encoder_read, encoder_written, _) = encoder
                                .encode_from_utf8(
                                    &intermediate_buffer[encoder_input_start..decoder_written],
                                    &mut output_buffer,
                                    last_output,
                                );
                            encoder_input_start += encoder_read;
                            if let Err(e) = write.write_all(&output_buffer[..encoder_written]) {
                                log::error!("Error writing output, error = {}", e);
                                return Err(DbcError::EncodingWriteOutputError);
                            }
                            match encoder_result {
                                CoderResult::InputEmpty => {
                                    break;
                                }
                                CoderResult::OutputFull => {
                                    continue;
                                }
                            }
                        }
                    }

                    // Now let's see if we should read again or process the
                    // rest of the current input buffer.
                    match decoder_result {
                        CoderResult::InputEmpty => {
                            break;
                        }
                        CoderResult::OutputFull => {
                            continue;
                        }
                    }
                }
            }
        }
    }

    Ok(())
}
