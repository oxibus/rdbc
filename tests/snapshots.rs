#![cfg(feature = "serde")]

use std::fs;
use std::path::Path;

use insta::{assert_debug_snapshot, assert_snapshot, assert_yaml_snapshot, with_settings};
use rrdbc::ast::network_ast::parse_dbc;
use rrdbc::encoding::decode_cp1252;

/// Test parsing all DBC files
#[test]
fn test_cantools_dbc_files() {
    let test_dirs = [
        ("dbc/dbc_from_cantools", "dbc-cantools"),
        ("dbc/mytest", "mytest"),
    ];
    for (path, file_name) in &test_dirs {
        let path = Path::new(path);
        let snapshot_path = String::from("snapshots-") + file_name;
        // snapshots should go to /tests/snapshots-* directory
        with_settings! {
            { omit_expression => true,
              snapshot_path => snapshot_path ,
              prepend_module_to_snapshot => false },
            {
                test_dbc_files(path);
            }
        }
    }
}

/// Test parsing all DBC files in the given directory.
fn test_dbc_files(dir: impl AsRef<Path>) {
    let dir = dir.as_ref();
    let dir_display = dir.display();
    let dir_content = fs::read_dir(dir).unwrap_or_else(|e| {
        panic!(
            "
--------------------------------------------------------------------------
Error reading dbc test files from   {dir_display}
{e}
Make sure git submodules are up to date by running
    git submodule update --init --recursive
--------------------------------------------------------------------------
"
        )
    });

    eprintln!("Testing dbc files in directory: {dir_display}");
    for dbc_path in dir_content {
        let path = dbc_path.unwrap().path();
        if path
            .extension()
            .is_some_and(|ext| ext.eq_ignore_ascii_case("dbc"))
        {
            parse_one_file(&path);
        }
    }
}

/// Parse a single DBC file and assert a snapshot of the result.
fn parse_one_file(path: &Path) {
    eprintln!("Testing DBC file: {}", path.display());
    let file_name = path.file_stem().unwrap().to_string_lossy().to_string();
    let buffer = fs::read(path).unwrap();
    if let Some(buffer) = decode_cp1252(&buffer) {
        match parse_dbc(buffer.as_ref()) {
            Ok(dbc) => assert_yaml_snapshot!(file_name, dbc),
            Err(e) => {
                eprintln!("Failed to parse {file_name}.dbc: {e:?}");
                assert_debug_snapshot!(format!("!error___{file_name}"), e);
            }
        }
    } else {
        let error = format!("Failed to decode {file_name}.dbc as cp1252");
        eprintln!("{error}");
        assert_snapshot!(format!("!error___{file_name}"), error);
    }
}
