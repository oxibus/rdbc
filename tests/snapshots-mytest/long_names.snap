---
source: tests/snapshots.rs
---
version: ""
new_symbols:
  - NS_DESC_
  - CM_
  - BA_DEF_
  - BA_
  - VAL_
  - CAT_DEF_
  - CAT_
  - FILTER
  - BA_DEF_DEF_
  - EV_DATA_
  - ENVVAR_DATA_
  - SGTYPE_
  - SGTYPE_VAL_
  - BA_DEF_SGTYPE_
  - BA_SGTYPE_
  - SIG_TYPE_REF_
  - VAL_TABLE_
  - SIG_GROUP_
  - SIG_VALTYPE_
  - SIGTYPE_VALTYPE_
  - BO_TX_BU_
  - BA_DEF_REL_
  - BA_REL_
  - BA_DEF_DEF_REL_
  - BU_SG_REL_
  - BU_EV_REL_
  - BU_BO_REL_
  - SG_MUL_VAL_
bit_timing:
  value: ~
nodes:
  - NN123456789012345678901234567890
  - N12345678901234567890123456_0000
  - N1234567890123456789012345678901
  - N12345678901234567890123456_0001
value_tables: ~
messages:
  - header:
      id: 9
      name: SS123456789012345678901234587890
      size: 8
      transmitter: Vector__XXX
    signals: []
  - header:
      id: 8
      name: SS1234567890123456789012345_0000
      size: 8
      transmitter: Vector__XXX
    signals: []
  - header:
      id: 7
      name: SS1234567890123456789012345_0001
      size: 8
      transmitter: Vector__XXX
    signals: []
  - header:
      id: 6
      name: SS123456789012345678901234577890
      size: 8
      transmitter: Vector__XXX
    signals: []
  - header:
      id: 5
      name: SS123456789012345678901234567890
      size: 8
      transmitter: Vector__XXX
    signals: []
  - header:
      id: 4
      name: S1234567890123456789012345678901
      size: 8
      transmitter: Vector__XXX
    signals:
      - name: SS123456789012345678901234567890
        multiplexer: ~
        start_bit: 0
        size: 8
        byte_order: LittleEndian
        value_type: Signed
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
  - header:
      id: 0
      name: M12345678901234567890123456_0000
      size: 8
      transmitter: Vector__XXX
    signals:
      - name: SSS12345678901234567890123456789
        multiplexer: ~
        start_bit: 0
        size: 8
        byte_order: LittleEndian
        value_type: Signed
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
  - header:
      id: 1
      name: M1234567890123456789012345678901
      size: 8
      transmitter: N1234567890123456789012345678901
    signals:
      - name: SS1234567890123456789012345_0000
        multiplexer: ~
        start_bit: 32
        size: 8
        byte_order: LittleEndian
        value_type: Signed
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
      - name: SS1234567890123456789012345_0001
        multiplexer: ~
        start_bit: 24
        size: 8
        byte_order: LittleEndian
        value_type: Signed
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
      - name: SS1234567890123456789012345_0002
        multiplexer: ~
        start_bit: 16
        size: 8
        byte_order: LittleEndian
        value_type: Signed
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - N12345678901234567890123456_0000
      - name: S12345678901234567890123456_0000
        multiplexer: ~
        start_bit: 8
        size: 8
        byte_order: LittleEndian
        value_type: Signed
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
      - name: S1234567890123456789012345678901
        multiplexer: ~
        start_bit: 0
        size: 8
        byte_order: LittleEndian
        value_type: Signed
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
  - header:
      id: 2
      name: M12345678901234567890123456_0001
      size: 8
      transmitter: N12345678901234567890123456_0001
    signals:
      - name: SS1234567890123456789012345_0003
        multiplexer: ~
        start_bit: 24
        size: 8
        byte_order: LittleEndian
        value_type: Signed
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
      - name: SS1234567890123456789012345_0004
        multiplexer: ~
        start_bit: 16
        size: 8
        byte_order: LittleEndian
        value_type: Signed
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
      - name: S12345678901234567890123456_0001
        multiplexer: ~
        start_bit: 8
        size: 8
        byte_order: LittleEndian
        value_type: Signed
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
      - name: S12345678901234567890123456_0002
        multiplexer: ~
        start_bit: 0
        size: 8
        byte_order: LittleEndian
        value_type: Signed
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
  - header:
      id: 3
      name: MM123456789012345678901234567890
      size: 8
      transmitter: Vector__XXX
    signals:
      - name: SSS123456789012345678901234_0000
        multiplexer: ~
        start_bit: 8
        size: 8
        byte_order: LittleEndian
        value_type: Signed
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
      - name: SS1234567890123456789012345_0005
        multiplexer: ~
        start_bit: 0
        size: 8
        byte_order: LittleEndian
        value_type: Signed
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
env_vars:
  - env_var_name: E12345678901234567890123456_0000
    env_var_type: Integer
    minimum: 0
    maximum: 0
    unit: ""
    initial_value: 0
    ev_id: 1
    access_type: 0
    access_nodes:
      - Vector__XXX
  - env_var_name: E1234567890123456789012345678901
    env_var_type: Integer
    minimum: 0
    maximum: 0
    unit: ""
    initial_value: 0
    ev_id: 2
    access_type: 0
    access_nodes:
      - Vector__XXX
env_vars_data: []
comments: []
attribute_definitions:
  - Network:
      attribute_name: BusType
      attribute_value_type:
        String: {}
  - Signal:
      attribute_name: SystemSignalLongSymbol
      attribute_value_type:
        String: {}
  - Message:
      attribute_name: SystemMessageLongSymbol
      attribute_value_type:
        String: {}
  - Node:
      attribute_name: SystemNodeLongSymbol
      attribute_value_type:
        String: {}
  - EnvironmentVariable:
      attribute_name: SystemEnvVarLongSymbol
      attribute_value_type:
        String: {}
attribute_defaults:
  - Attribute:
      attribute_name: BusType
      attribute_value:
        String: ""
  - Attribute:
      attribute_name: SystemSignalLongSymbol
      attribute_value:
        String: ""
  - Attribute:
      attribute_name: SystemMessageLongSymbol
      attribute_value:
        String: ""
  - Attribute:
      attribute_name: SystemNodeLongSymbol
      attribute_value:
        String: ""
  - Attribute:
      attribute_name: SystemEnvVarLongSymbol
      attribute_value:
        String: ""
attribute_values:
  - Node:
      attribute_name: SystemNodeLongSymbol
      node_name: NN123456789012345678901234567890
      attribute_value:
        String: NN123456789012345678901234567890123
  - Node:
      attribute_name: SystemNodeLongSymbol
      node_name: N12345678901234567890123456_0000
      attribute_value:
        String: N123456789012345678901234567890123
  - Node:
      attribute_name: SystemNodeLongSymbol
      node_name: N12345678901234567890123456_0001
      attribute_value:
        String: N12345678901234567890123456789012
  - Message:
      attribute_name: SystemMessageLongSymbol
      message_id: 9
      attribute_value:
        String: SS12345678901234567890123458789012345
  - Message:
      attribute_name: SystemMessageLongSymbol
      message_id: 8
      attribute_value:
        String: SS1234567890123456789012345778901
  - Message:
      attribute_name: SystemMessageLongSymbol
      message_id: 7
      attribute_value:
        String: SS1234567890123456789012345878901234
  - Message:
      attribute_name: SystemMessageLongSymbol
      message_id: 5
      attribute_value:
        String: SS12345678901234567890123456789012
  - Message:
      attribute_name: SystemMessageLongSymbol
      message_id: 4
      attribute_value:
        String: S12345678901234567890123456789012
  - Message:
      attribute_name: SystemMessageLongSymbol
      message_id: 0
      attribute_value:
        String: M123456789012345678901234567890123
  - Message:
      attribute_name: SystemMessageLongSymbol
      message_id: 2
      attribute_value:
        String: M12345678901234567890123456789012
  - Message:
      attribute_name: SystemMessageLongSymbol
      message_id: 3
      attribute_value:
        String: MM12345678901234567890123456789012
  - Signal:
      attribute_name: SystemSignalLongSymbol
      message_id: 4
      signal_name: SS123456789012345678901234567890
      attribute_value:
        String: SS12345678901234567890123456789012
  - Signal:
      attribute_name: SystemSignalLongSymbol
      message_id: 0
      signal_name: SSS12345678901234567890123456789
      attribute_value:
        String: SSS12345678901234567890123456789012
  - Signal:
      attribute_name: SystemSignalLongSymbol
      message_id: 1
      signal_name: SS1234567890123456789012345_0000
      attribute_value:
        String: SS12345678901234567890123456789012332
  - Signal:
      attribute_name: SystemSignalLongSymbol
      message_id: 1
      signal_name: SS1234567890123456789012345_0001
      attribute_value:
        String: SS1234567890123456789012345678901233
  - Signal:
      attribute_name: SystemSignalLongSymbol
      message_id: 1
      signal_name: SS1234567890123456789012345_0002
      attribute_value:
        String: SS12345678901234567890123456789012
  - Signal:
      attribute_name: SystemSignalLongSymbol
      message_id: 1
      signal_name: S12345678901234567890123456_0000
      attribute_value:
        String: S123456789012345678901234567890123
  - Signal:
      attribute_name: SystemSignalLongSymbol
      message_id: 2
      signal_name: SS1234567890123456789012345_0003
      attribute_value:
        String: SS12345678901234567890123456789012dw
  - Signal:
      attribute_name: SystemSignalLongSymbol
      message_id: 2
      signal_name: SS1234567890123456789012345_0004
      attribute_value:
        String: SS12345678901234567890123456789012
  - Signal:
      attribute_name: SystemSignalLongSymbol
      message_id: 2
      signal_name: S12345678901234567890123456_0001
      attribute_value:
        String: S12345678901234567890123456789012
  - Signal:
      attribute_name: SystemSignalLongSymbol
      message_id: 2
      signal_name: S12345678901234567890123456_0002
      attribute_value:
        String: S123456789012345678901234567890123
  - Signal:
      attribute_name: SystemSignalLongSymbol
      message_id: 3
      signal_name: SSS123456789012345678901234_0000
      attribute_value:
        String: SSS12345678901234567890123456789012
  - Signal:
      attribute_name: SystemSignalLongSymbol
      message_id: 3
      signal_name: SS1234567890123456789012345_0005
      attribute_value:
        String: SS1234567890123456789012345678901233
  - EnvironmentVariable:
      attribute_name: SystemEnvVarLongSymbol
      env_var_name: E12345678901234567890123456_0000
      attribute_value:
        String: E12345678901234567890123456789012
signal_value_descriptions: []
env_var_value_descriptions: []
